client/build_test.go

```diff
@@ -175,3 +184,367 @@ func testUnknownBuildID(t *testing.T, sb integration.Sandbox) {
+			Cwd:    "/",
+			Tty:    false,
+			Stdin:  ioutil.NopCloser(stdin2),
+			Stdout: stdout2,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		// next part is: `| cat > /tmp/test`
+		pid3, err := ctr.Start(ctx, client.StartRequest{
+			Args:  []string{"sh", "-c", "cat > /tmp/test"},
+			Cwd:   "/",
+			Stdin: stdin3,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		err = pid2.Wait()
+		if err != nil {
+			stdout2.Close()
+			return nil, err
+		}
+
+		err = stdout2.Close()
+		if err != nil {
+			return nil, err
+		}
+
+		err = pid3.Wait()
+		if err != nil {
+			return nil, err
+		}
+
+		err = stdin3.Close()
+		if err != nil {
+			return nil, err
+		}
+
+		pid4, err := ctr.Start(ctx, client.StartRequest{
+			Args:   []string{"cat", "/tmp/test"},
+			Cwd:    "/",
+			Stdout: &nopCloser{output},
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		err = pid4.Wait()
+		if err != nil {
+			return nil, err
+		}
+
+		return &client.Result{}, nil
+	}
+
+	_, err = c.Build(ctx, SolveOpt{}, product, b, nil)
+	require.NoError(t, err)
+	require.Equal(t, "testing", output.String())
+
+	checkAllReleasable(t, c, sb, true)
+}
+
+// testClientGatewayContainerPID1Fail is testing clean shutdown and release
+// of resources when the primary pid1 exits with non-zero exit status
+func testClientGatewayContainerPID1Fail(t *testing.T, sb integration.Sandbox) {
+	requiresLinux(t)
+
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	product := "buildkit_test"
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		st := llb.Image("busybox:latest")
+
+		def, err := st.Marshal(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal state")
+		}
+
+		r, err := c.Solve(ctx, client.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to solve")
+		}
+
+		ctr, err := c.NewContainer(ctx, client.NewContainerRequest{
+			Mounts: []client.Mount{{
+				Dest:      "/",
+				MountType: pb.MountType_BIND,
+				Ref:       r.Ref,
+			}},
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		pid1, err := ctr.Start(ctx, client.StartRequest{
+			Args: []string{"false"},
+			Cwd:  "/",
+		})
+		if err != nil {
+			ctr.Release(ctx)
+			return nil, err
+		}
+
+		defer ctr.Release(ctx)
+		err = pid1.Wait()
+
+		var exitError *errdefs.ExitError
+		require.True(t, errors.As(err, &exitError))
+		require.Equal(t, uint32(1), exitError.ExitCode)
+
+		return nil, err
+	}
+
+	_, err = c.Build(ctx, SolveOpt{}, product, b, nil)
+	require.Error(t, err)
+
+	checkAllReleasable(t, c, sb, true)
+}
+
+// testClientGatewayContainerPID1Exit is testing that all process started
+// via `Exec` are shutdown when the primary pid1 process exits
+func testClientGatewayContainerPID1Exit(t *testing.T, sb integration.Sandbox) {
+	if sb.Rootless() {
+		// TODO fix this
+		// We get `panic: cannot statfs cgroup root` when running this test
```

Comment:
Could you open an issue in runc repo

Reply:
Why does it appear in exec and no in regular builds?

Reply:
Created issue, with stacktrace from runc: https://github.com/opencontainers/runc/issues/2573

Reply:
`/sys/fs/cgroup` isn't mounted?

Reply:
It looks like the `/sys` mounts are trimmed out [here](https://github.com/moby/buildkit/blob/master/util/rootless/specconv/specconv_linux.go#L29)

Reply:
It appears to work if we bind mount in `/sys`, but not sure is this is the right thing to do here.  After diffing the `config.json` from `runc spec` and `runc spec --rootless`.  I added this to that `ToRootless` function and it worked for my tests, but not sure if there are other complications:
```
	spec.Mounts = append(spec.Mounts, specs.Mount{
		Destination: "/sys",
		Type:        "none",
		Source:      "/sys",
		Options:     []string{"rbind", "nosuid", "noexec", "nodev", "ro"},
	})
```
Does this seem like the right solution?

Reply:
Sorry I didn't notice this comment.

`{"rbind", "ro"}` doesn't make submounts recursively read-only and can result in container breakout.
So we should mount `/sys` and `/sys/fs/cgroup` separately as `{"bind", "ro"}`.

In the long term, we should fix https://github.com/opencontainers/runc/issues/2573

Reply:
Okay, I have updated with this: https://github.com/moby/buildkit/pull/1627/commits/4b51fbd48c461df30e5b8e078bdeed8858aa4099

It seems to work, hopefully does not cause other issues.

Reply:
This change ended up causing several other integration tests to fail with oci-rootless, like: `TestIntegration/TestDefaultEnvWithArgs` so I am going to revert it.  I am leaning towards leaving my "skip rootless" hack in place for this PR and then investigating/fixing the rootless + /sys issue later.